cdlang/src/main/java/de/monticore/cd/codegen/decorators/DeepCloneAndDeepEqualsDecorator.java

```diff
@@ -0,0 +1,359 @@
+ * When cloning an object, we first check if the object is already in the map.<br>
+ * If we encounter an object we have not seen yet, we create a new one and add it to our map.
+ * This is crucial because if that object later contains a reference to itself (either directly or
+ * indirectly),
+ * we will recognize it from our map.
+ * This prevents us from getting stuck in an endless loop trying to create the same object over and
+ * over;
+ * instead, we just copy the existing reference found in the map<br>
+ * The map is a Map‹Object, Object› where the key is the original object and the value is the copied
+ * object of the key.
+ * If we do not pass the map, we would end up with a stack overflow error when trying to clone
+ * cyclic references.<br>
+ * <br>
+ * <br>
+ * The deepEquals method is also three methods: <br>
+ * 1. deepEquals(o: Object): <br>
+ * This method calls the deepEquals method with the signature deepEquals(o: Object,
+ * <code>forceSameOrder</code>: boolean).<br>
+ * <br>
+ * 2. deepEquals(o: Object, <code>forceSameOrder</code>: boolean): <br>
+ * This method calls the deepEquals method with the signature deepEquals(o: Object,
+ * <code>forceSameOrder</code>: boolean, visitedObjects: Set<Object>).
+ * With a new Map‹Object, Set‹Object›› of visited objects to avoid cyclic references. <br>
+ * <br>
+ * 3. deepEquals(o: Object, <code>forceSameOrder</code>: boolean, <code>visitedObjects</code>:
+ * Set‹Object›): <br>
+ * This method is the actual implementation of the deepEquals method.<br>
+ * It compares the object with the current instance and checks if the attributes are equal.<br>
+ * It begins by adding the <code>currentObject</code> to the map of <code>visitedObjects</code>.<br>
+ * The map maps objects found in the first object onto objects found for that specific object in the
+ * second object.<br>
+ * Then it resolves the <code>currentObject</code>.<br>
+ * Because we added the <code>currentObject</code> to the set of <code>visitedObjects</code> of the
+ * specific first object,
+ * we can detect cyclic references and avoid them.<br>
+ * Afterward, we remove the <code>currentObject</code> from the map of <code>visitedObjects</code>
+ * to allow for further comparisons.<br>
+ * TODO currently the deepEquals method is not symmetric, meaning that if A.equals(B) is true,
+ * B.equals(A) is not necessarily true.
+ */
+public class DeepCloneAndDeepEqualsDecorator extends AbstractDecorator<AbstractDecorator.NoData>
+    implements CDBasisVisitor2 {
+  
+  /**
+   * a collection of all classes from the class diagram as strings
+   */
+  List<String> classesFromClassdiagramAsString = new ArrayList<>();
+  boolean isInitialized = false;
+  
+  @Override
+  @SuppressWarnings("rawtypes")
+  public Iterable<Class<? extends IDecorator>> getMustRunAfter() { return super.getMustRunAfter(); }
+  
+  private void initClassesFromClassDiagramAsString(ASTNode node) {
```

Comment:
```suggestion
  protected void initClassesFromClassDiagramAsString(ASTNode node) {
```

Try to avoid `private` (to allow extending the class)